src/content/docs/en/basics/layouts.mdx

```diff
@@ -258,6 +258,28 @@ const { title } = Astro.props.frontmatter || Astro.props;
 </html>
 ```
 
+### Working with TypeScript
+
+Any astro layout can be modified to introduce typesafety & autocompletion via
+
+```astro ins={3} ": CustomProps"
+---
+// src/components/MyLayout.astro
+type CustomProps = { title: string };
+const { title } : CustomProps = Astro.props.frontmatter || Astro.props;
+---
+```
+
+or by overriding default Props via
+
+```astro ins={3} 
+---
+// src/components/MyLayout.astro
+interface Props { title: string }
+const { title } = Astro.props;
+---
+```
```

Comment:
``````suggestion
Any astro layout can be modified to introduce typesafety & autocompletion by providing the types for your props:

```astro ins={2} title="src/components/MyLayout.astro"
---
interface Props { 
	title: string
}
const { title } = Astro.props;
---
```

You can override Astro's default `Props` to handle situations where your layout recieves props of different types:

```astro ": CustomProps" title="src/components/MyLayout.astro"
---
type CustomProps = { title: string };
const { title } : CustomProps = Astro.props.frontmatter || Astro.props;
---
```
``````
So here's what I was thinking...

- Showing the default using TypeScript first feels more natural, like "this is how you type your props" before a more niche/advanced case like overriding the props.
- I do like how compact the code sample is, but I think whenever I've seen this actually used in someone's project, the `interfaceProps` is all expanded. So I'm thinking maybe we show it like that? Someone might often have more than one prop, and this does help keep their code readable, too.
- each of these need some kind of actual explanation of what's happening (not just "via", which is like "like so" or "like this" and makes the reader guess which part of the next code is the part you mean). I've tried to explain with words I think make sense, but I will need someone more familiar with TypeScript terminology to verify!

Reply:
- that makes sense, I went autopilot with this one just being worried for the whole `Astro.props.frontmatter` issue getting addressed with high priority
- Sure, I will look into this. There are exemplar props I can/will refer from https://docs.astro.build/en/basics/layouts/#markdown-layout-props
- that helps the situation significantly 🙌 I will be `ptal`ing this to get accurate info

Reply:
I was thinking like our example packages show the props like this:

![image](https://github.com/withastro/docs/assets/5098874/f656ebd1-c952-42ce-b190-173c530bcbe8)